connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+    # TODO: improve utils.retryable to allow custom logic
+    # that can help choose what to retry
+    def wrapper(func):
+        @wraps(func)
+        async def wrapped(*args, **kwargs):
+            retry = 1
+            while retry <= retries:
+                try:
+                    async for item in func(*args, **kwargs):
+                        yield item
+                    break
+                except Exception as e:
+                    if retry >= retries:
+                        raise e
+
+                    retry += 1
+
+        return wrapped
+
+    return wrapper
+
+
+class ThrottledError(Exception):
+    """Internal exception class to indicate that request was throttled by the API"""
+
+    pass
+
+
+class SlackClient:
+    def __init__(self, configuration):
+        self.token = configuration["token"]
+        self._http_session = aiohttp.ClientSession(
+            headers=self._headers(),
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._sleeps = CancellableSleeps()
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    def ping(self):
+        return
+
+    def close(self):
+        self._http_session.close()
+        return
+
+    async def list_channels(self, only_my_channels, cursor=None):
+        conversation_batch_limit = 200
+        self._logger.debug("Iterating over all channels")
+        if only_my_channels:
+            self._logger.debug("Will only yield channels the bot belongs to")
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[LIST_CONVERSATIONS]}?limit={conversation_batch_limit}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+
+            for channel in response["channels"]:
+                if only_my_channels and not channel.get("is_member", False):
+                    continue
+                self._logger.debug(f"Yielding channel '{channel['name']}'")
+                yield channel
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def join_channel(self, channel_id):
+        url = f"{BASE_URL}/{ENDPOINTS[JOIN_CONVERSATIONS]}?channel={channel_id}"
+        return await self._get_json(url)
+
+    async def list_messages(self, channel_id, oldest, latest):
+        self._logger.info(
+            f"Fetching messages between {oldest} and {latest} from channel: '{channel_id}'"
+        )
+        message_batch_limit = 200
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[CONVERSATION_HISTORY]}?channel={channel_id}&limit={message_batch_limit}&oldest={oldest}&latest={latest}"
+            response = await self._get_json(url)
+            for message in response.get("messages", []):
+                latest = message.get("ts")
+                if message.get("type") == "message":
+                    if message.get("reply_count", 0) > 0:
```

Comment:
Hmmm. what about this `retry_count`? What's that logic? Curious!

Reply:
I'm confused. I _think_ you were noting that I stole your retry logic from Sharepoint (which I did, shamelessly. And not just that logic). But this is _reply_ count. 

This was the indicator I chose for if a given message is the parent for a thread. If it has replies, then you have to make a follow-up request(s) to get all those replies, since they're not part of `conversation.history`

Reply:
Oh wow, I read it as a "retry", my bad 😮‍💨

Reply:
You are not alone.